ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/datacolumns/util/DataColumnSidecarELRecoveryManagerImpl.java

```diff
@@ -0,0 +1,371 @@
+import tech.pegasys.teku.spec.datastructures.networking.libp2p.rpc.BlobIdentifier;
+import tech.pegasys.teku.spec.datastructures.type.SszKZGCommitment;
+import tech.pegasys.teku.spec.executionlayer.ExecutionLayerChannel;
+import tech.pegasys.teku.spec.logic.versions.deneb.helpers.MiscHelpersDeneb;
+import tech.pegasys.teku.spec.logic.versions.deneb.types.VersionedHash;
+import tech.pegasys.teku.spec.logic.versions.fulu.helpers.MiscHelpersFulu;
+import tech.pegasys.teku.statetransition.blobs.RemoteOrigin;
+import tech.pegasys.teku.statetransition.datacolumns.CustodyGroupCountManager;
+import tech.pegasys.teku.statetransition.datacolumns.DataColumnSidecarELRecoveryManager;
+import tech.pegasys.teku.statetransition.util.AbstractIgnoringFutureHistoricalSlot;
+import tech.pegasys.teku.storage.client.RecentChainData;
+
+public class DataColumnSidecarELRecoveryManagerImpl extends AbstractIgnoringFutureHistoricalSlot
+    implements DataColumnSidecarELRecoveryManager {
+  private static final Logger LOG = LogManager.getLogger();
+
+  private final Map<Bytes32, RecoveryTask> recoveryTasks = new HashMap<>();
+  private final NavigableSet<SlotAndBlockRoot> orderedRecoveryTasks = new TreeSet<>();
+  private final Spec spec;
+  private final AsyncRunner asyncRunner;
+  private final RecentChainData recentChainData;
+  private final ExecutionLayerChannel executionLayer;
+  private final int maxTrackers;
+  private final Consumer<List<DataColumnSidecar>> dataColumnSidecarPublisher;
+  private final CustodyGroupCountManager custodyGroupCountManager;
+  private final KZG kzg;
+
+  private final Supplier<MiscHelpersFulu> miscHelpersFuluSupplier;
+
+  @VisibleForTesting
+  public DataColumnSidecarELRecoveryManagerImpl(
+      final Spec spec,
+      final AsyncRunner asyncRunner,
+      final RecentChainData recentChainData,
+      final ExecutionLayerChannel executionLayer,
+      final UInt64 historicalSlotTolerance,
+      final UInt64 futureSlotTolerance,
+      final int maxTrackers,
+      final KZG kzg,
+      final Consumer<List<DataColumnSidecar>> dataColumnSidecarPublisher,
+      final CustodyGroupCountManager custodyGroupCountManager) {
+    super(spec, futureSlotTolerance, historicalSlotTolerance);
+    this.spec = spec;
+    this.asyncRunner = asyncRunner;
+    this.recentChainData = recentChainData;
+    this.executionLayer = executionLayer;
+    this.maxTrackers = maxTrackers;
+    this.kzg = kzg;
+    this.dataColumnSidecarPublisher = dataColumnSidecarPublisher;
+    this.custodyGroupCountManager = custodyGroupCountManager;
+    this.miscHelpersFuluSupplier =
+        () -> MiscHelpersFulu.required(spec.forMilestone(SpecMilestone.FULU).miscHelpers());
+  }
+
+  @Override
+  public void onNewDataColumnSidecar(
+      final DataColumnSidecar dataColumnSidecar, final RemoteOrigin remoteOrigin) {
+    if (!spec.atSlot(dataColumnSidecar.getSlot())
+        .getMilestone()
+        .isGreaterThanOrEqualTo(SpecMilestone.FULU)) {
```

Comment:
```suggestion
    if (spec.atSlot(dataColumnSidecar.getSlot())
        .getMilestone()
        .isLessThan(SpecMilestone.FULU)) {
```

Reply:
Q: Is there a scenario where we receive a data column sidecar before we are in Fulu? My understanding is that we are only subscribing to the DataColumnSidecar subnets when we are in Fulu.

I think some of these cases where we are early returning from the method it is worth adding a debug log. Otherwise we will go crazy when testing this thing.